### PR TITLE
Include existing Source Maps when passing options to Autoprefixer / P…

### PR DESCRIPTION
**Pull request by npetruzzelli**
…ostCSS.

I'm new to the pull requests and contributing to the codebase of others. Please let me know if I've left out expected information or missed a common developer courtesy. My intent is to provide information about how I arrived at proposing this change.
# TL;DR

For a few builds I've had some issues with CSS sourcemaps not pointing to the correct SASS files when `gulp-autoprefixer` is used. `gulp-sass` is being used in this case, but initial tests with `gulp-ruby-sass`, `gulp-less`, `gulp-myth`, and `gulp-stylus` seemed to be consistent. If problems persist, I'll go back and do more thorough checks with them.

A small change will generate correct sourcemaps when they already exist on a given file. It has to do with passing existing sourcemaps to PostCSS.

[Line 21 in `gulp-autoprefixer`](https://github.com/sindresorhus/gulp-autoprefixer/blob/v3.1.0/index.js#L21):

``` diff
--  map: file.sourceMap ? {annotation: false} : false,
++  map: file.sourceMap ? {annotation: false, prev: file.sourceMap} : false,
```
# Details (the long version)
## The ByScript "fix"

There were a couple work-arounds out there, but the one that seemed to return frequently in my search is an issue in the `gulp-sourcemaps` repository by @floridoo . https://github.com/floridoo/gulp-sourcemaps/issues/60

The user @ByScripts came up with the solution to write inline source map comments and then load them again. The modification to your task looks similar to the following:

``` javascript
    .pipe($.sourcemaps.write({includeContent: false}))
    .pipe($.sourcemaps.init({loadMaps: true}))
```

This doesn't work as you'd expect it to. The original comment never gets cleaned up by `convert-source-map` (https://github.com/floridoo/gulp-sourcemaps/blob/v1.6.0/index.js#L41-L49) because the files that are being worked with still have a `file.sourceMap` property (https://github.com/floridoo/gulp-sourcemaps/blob/v1.6.0/index.js#L20-L23).

The reason it appeared to work is the original inline comment still existed in the file and that is what the browser was picking up, but it was never being updated appropriately as it moved down the line and would lead to multiple `sourceMappingURL` comments. To confirm this, a small local module can be used:

``` javascript

// THE PLUGIN

const through = require('through2');

module.exports = function removeSourcemapProperty(){
    function removeGulpSourcemapProperty(file, encoding, callback){
        if (file.sourceMap) {
            delete file.sourceMap;
        }
        this.push(file);
        return callback();
    }
    return through.obj(removeGulpSourcemapProperty);
}

// THE PIPES

    .pipe($.sourcemaps.write({includeContent: false}))
    .pipe(removeSourcemapProperty())
    .pipe($.sourcemaps.init({loadMaps: true}))
```

The comment gets correctly removed and the sourcemaps are back to (correctly?) being broken.
## The _source_ of the problem.

The next step was to log the state of `file.sourceMap` and related variables as it passed through `gulp-sourcemaps` and `gulp-autoprefixer`. I arrived here: https://github.com/sindresorhus/gulp-autoprefixer/blob/v3.1.0/index.js#L21. My search brought me directly to the `autoprefixer` and `postcss` repositories where I found the following document: https://github.com/postcss/postcss/blob/5.1.0/docs/source-maps.md.

This lead to adding the `prev` property in Line 21 in `gulp-autoprefixer`:

``` diff
--  map: file.sourceMap ? {annotation: false} : false,
++  map: file.sourceMap ? {annotation: false, prev: file.sourceMap} : false,
```

And everything now appears to work as expected.

### CODE REVIEW

**Comment by npetruzzelli** [comment]
Apologies, it seems I missed due diligence. This appears to be related to: https://github.com/sindresorhus/gulp-autoprefixer/issues/55

**Comment by npetruzzelli** [comment]
After finding my way to <https://github.com/sindresorhus/gulp-autoprefixer/pull/61#discussion_r55338955>, am I correct in saying that you beleive it is better that `PostCSS` handles `gulp-sourcemaps` support since it would instantly benefit all consumers of `PostCSS`? Not just `Autoprefixer`, and by extension, `gulp-autoprefixer`.

I can definitely appreciate and see the merit in that. If that is the case, I would counter that argument with "at what point does `PostCSS` stop favoring a specific third party tool?" be it `gulp-sourcemaps`, or some other source map plugin, perhaps for another system like Grunt?

Supplying the `prev` property allows us to support existing source maps without `PostCSS` getting drawn in to catering to specific libraries, which could inflate their code base with edge cases and make it more difficult to maintain.

And to counter my own counter argument, solving the problem in `PostCSS` might look something like this:

Line 134 in <https://github.com/postcss/postcss/blob/5.1.0/lib/previous-map.es6#L98-L134>

```diff
      loadMap(file, prev) {
          if ( prev === false ) return false;
  
          if ( prev ) {
              if ( typeof prev === 'string' ) {
                  return prev;
              } else if ( typeof prev === 'function' ) {
                  let prevPath = prev(file);
                  if ( prevPath && fs.existsSync && fs.existsSync(prevPath) ) {
                      return fs.readFileSync(prevPath, 'utf-8').toString().trim();
                  } else {
                      throw new Error('Unable to load previous source map: ' +
                      prevPath.toString());
                  }
              } else if ( prev instanceof mozilla.SourceMapConsumer ) {
                  return mozilla.SourceMapGenerator
                      .fromSourceMap(prev).toString();
              } else if ( prev instanceof mozilla.SourceMapGenerator ) {
                  return prev.toString();
              } else if ( this.isMap(prev) ) {
                  return JSON.stringify(prev);
              } else {
                  throw new Error('Unsupported previous source map format: ' +
                      prev.toString());
              }
  
          } else if ( this.inline ) {
              return this.decodeInline(this.annotation);
  
          } else if ( this.annotation ) {
              let map = this.annotation;
              if ( file ) map = path.join(path.dirname(file), map);
  
              this.root = path.dirname(map);
              if ( fs.existsSync && fs.existsSync(map) ) {
                  return fs.readFileSync(map, 'utf-8').toString().trim();
              } else {
                  return false;
              }
++        } else if ( file.sourceMap ) {
++            // QUESTION: Where should file.sourceMap checking happen? first? last?
++            // 
++            // After false check. False means the consumer explicitly wants to ignore previously existing source maps.
++            // After the prev check. This is something explicitly passed by the consumer and should not be overridden.
++            // So the question that remains is: before `this.inline`? before `this.annotation`? or after `this.annotation`?
++            // 
++            // What happens if there is both an annotation and a `file.sourceMap` property? Someone didn't initialize `gulp-sourcemaps` correctly. They should have used the `{loadMaps:true}` option. This converts existing maps (inline or external) to the `file.sourceMap` property.
++            // It is not the job of `PostCSS` to resolve this conflict, but the order in which it is checked may imply that it favors one over the other.
++            // 
++            // SOLUTION: after `this.annotation`
++            // This preserves the existing order/behavior of `PostCSS`. It encourages/requires consumers of `gulp-sourcemaps` to understand more about the initialization process.
++            if ( this.isMap(file.sourceMap) ) {
++                return JSON.stringify(file.sourceMap);
++            } else {
++                return false; //Throw error instead?
++            }
          }
      }
```

What do you think?

**Comment by sindresorhus** [comment]
This plugin [already passes the previous source map and its own](
https://github.com/sindresorhus/gulp-autoprefixer/blob/324ff05ffc9e45211a40f7aa2ac5127075378a18/index.js#L28) into [`vinyl-sourcemaps-apply`](https://github.com/floridoo/vinyl-sourcemaps-apply) which is supposed to correctly merge them. If that doesn't work, you should take it up on the `vinyl-sourcemaps-apply` issue tracker. It might also be an issue with the Source Map coming from PostCSS. But this plugin implements it correctly, like all my other gulp plugins, which does not inhibit these issues.

**Comment by npetruzzelli** [comment]
Thanks Sindre. I have some additional details. 

# 1) I need to correct myself

My counter argument above regarding making a fix in PostCSS's `previous-map.es6` file is incorrect. `file` represents the `from` property, so it is a string and wouldn't contain any information from the file object itself. If that wasn't the case, a minor change would need to be made to `if (prev === false) return false;` for any of my suggestion to work at all. At no point does `Autoprefixer` or `PostCSS` have access to the file object. I was wrong in that argument/hypothetical solution.

# 2) The problem doesn't happen for files at the top level

**TL;DR** I'll dig deeper into `vinyl-sourcemaps-apply` and open an issue in that repository that points to this pull request as a reference.

Consider the following is the project folder structure.

```
source/
    sub-directory/
        app.scss
    top.scss
gulpfile.js
package.json
```

`npm i --save-dev gulp gulp-autoprefixer gulp-load-plugins gulp-plumber gulp-sass gulp-sourcemaps`

```json
{
  "private":true,
  "name": "test-sourcemaps",
  "version": "1.0.0",
  "devDependencies": {
    "gulp": "^3.9.1",
    "gulp-autoprefixer": "^3.1.0",
    "gulp-load-plugins": "^1.2.4",
    "gulp-plumber": "^1.1.0",
    "gulp-sass": "^2.3.2",
    "gulp-sourcemaps": "^1.6.0"
  }
}
```


```javascript
/* ./gulpfile.js */
const gulp            = require('gulp')
    , gulpLoadPlugins = require('gulp-load-plugins')
    , $               = gulpLoadPlugins()
;

gulp.task('default', (callback)=>{
    return gulp.src('source/**/*.scss', {
            base: 'source'
        })
        .pipe($.sourcemaps.init())
        .pipe($.sass.sync({
            outputStyle: 'expanded',
            precision: 10,
            includePaths: ['.']
        }).on('error', $.sass.logError))
        .pipe($.autoprefixer())
        .pipe($.sourcemaps.write('.', {
            includeContent: false
        }))
        .pipe(gulp.dest('dist/'))
    ;
});
```

```sass
/* ./source/top.scss */
$bgColor: cyan;

body{
    background:$bgColor;
}
```

```sass
/* ./source/sub-directory/app.scss */
$bgColor: lime;

body{
    background:$bgColor;
}
```
Modify `gulp-autoprefixer` to track it: <https://github.com/sindresorhus/gulp-autoprefixer/blob/v3.1.0/index.js>

```diff
          if (file.isStream()) {
              cb(new gutil.PluginError('gulp-autoprefixer', 'Streaming not supported'));
              return;
          }
  
          postcss(autoprefixer(opts)).process(file.contents.toString(), {
              map: file.sourceMap ? {annotation: false} : false,
              from: file.path,
              to: file.path
          }).then(function (res) {
              file.contents = new Buffer(res.css);
  
++            console.log('GULP-AUTOPREFIXER #1');
++            console.log(file.sourceMap);
              if (res.map && file.sourceMap) {
                  applySourceMap(file, res.map.toString());
              }
++            console.log('GULP-AUTOPREFIXER #2');
++            console.log(file.sourceMap);
```

The console will display the following: (using diff for convenient highlighting.)

```diff
  GULP-AUTOPREFIXER #1
  { version: 3,
    file: 'top.css',
    sources: [ 'top.scss' ],
    sourcesContent: [ '$bgColor : cyan;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ],
    mappings: 'AAEA,AAAA,IAAI,CAAA;EACA,UAAU,EAHH,IAAI;CAId',
    names: [] }
  GULP-AUTOPREFIXER #2
  { version: 3,
    sources: [ 'top.scss' ],
    names: [],
    mappings: 'AAEA;EACI,iBAHW;CAId',
    file: 'top.css',
    sourcesContent: [ '$bgColor : cyan;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ] }
  GULP-AUTOPREFIXER #1
  { version: 3,
    file: 'sub-directory/app.css',
++  sources: [ 'sub-directory/app.scss' ],
++  sourcesContent: [ '$bgColor : lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ],
    mappings: 'AAEA,AAAA,IAAI,CAAA;EACA,UAAU,EAHH,IAAI;CAId',
    names: [] }
  GULP-AUTOPREFIXER #2
  { version: 3,
--  sources: [ 'app.css' ],
    names: [],
    mappings: 'AAAA;EACE,iBAAiB;CAClB',
    file: 'app.css',
--  sourcesContent: [ 'body {\n  background: lime;\n}\n' ] }

```

We get the following:

**top.css.map**

```json
{"version":3,"sources":["top.scss"],"names":[],"mappings":"AAEA;EACI,iBAHW;CAId","file":"top.css"}
```

**app.css.map**

```json
{"version":3,"sources":["app.css"],"names":[],"mappings":"AAAA;EACE,iBAAiB;CAClB","file":"sub-directory/app.css"}
```

`top.css.map` still points to the SASS file, while `app.css.map` thinks that the output from `Autoprefixer` is the source.

I'm not at the point where I can explain why, but everything _just works_ when providing the `prev` option to `PostCSS`. @floridoo may have an idea of whats going on after I open an issue and point him to this PR.

Would it be alright to leave this PR open until it is resolved?

**Comment by npetruzzelli** [comment]
Issue opened: <https://github.com/floridoo/vinyl-sourcemaps-apply/issues/12>

**Comment by Maximaximum** [comment]
Hello guys! I'm quite new to sass and gulp, but I'm having this issue, too. I'd really appreciate it if you could fix it in any way. Thanks!

**Comment by npetruzzelli** [comment]
Unfortunately, I did not hear from the author of the module nor was I able to identify the root of the problem or a solution that differs from that in this pull request without changing to a different module.

Switching from `gulp-autoprefixer` to a combination of `gulp-postcss` and `autoprefixer` worked for me.

@Maximaximum : It may be worth looking into if your problem is the same as mine.

@sindresorhus : Thank you for your time and patience. I wish I could have come back with something more constructive.

**Comment by guohuihot** [comment]
.pipe($.sourcemaps.write({includeContent: false}))
    .pipe($.sourcemaps.init({loadMaps: true}))

**Comment by guohuihot** [comment]
gulp-autoprefixer
```
{"version":3,"sources":["../../src/css/demo.scss","../../src/css/_clearfix.scss","../../src/css/_text-overflow.scss","../../src/css/_opacity.scss","../../src/css/_transparent.scss","../../src/css/_nav-divider.scss","../../src/css/_font-demo.scss","../../src/css/_img-demo.scss","../../src/css/_modal.scss","../../src/css/_font-modal.scss","../../src/css/_variables.scss"],"names":[],"mappings":"","file":"demo.css","sourceRoot":"../../../src/css"}
```

no gulp-autoprefixer
```
{"version":3,"file":"demo.css","sources":["../../src/css/demo.scss","../../src/css/_ats.scss","../../src/css/_variables.scss","../../src/css/_clearfix.scss","../../src/css/_outline.scss","../../src/css/_text-overflow.scss","../../src/css/_opacity.scss","../../src/css/_transparent.scss","../../src/css/_triangle.scss","../../src/css/_nav-divider.scss","../../src/css/_reset-filter.scss","../../src/css/_font-demo.scss","../../src/css/_img-demo.scss","../../src/css/_modal.scss","../../src/css/_font-modal.scss"],"mappings":"","names":[],"sourceRoot":"../../../src/css"}
```
